File: 01_week2/fibonacci_sum_last_digit/fibonacci_sum_last_digit.py
# Task. Given an integer n, find the last digit of the sum F 0 + F 1 + · · · + F n .
# Uses python3
import sys

def fibonacci_sum_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum      = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current

    return sum % 10

def Fibonaccisumfast(n):
	n = (n+2)%60	
	if n<=1:
		return n
	previous=1
	beforePrevious=0
	current=-1
	for _ in range(2, n+1):
		current = (previous+beforePrevious)%10
		beforePrevious = previous
		previous = current
	return current-1


if __name__ == '__main__':
    #input = sys.stdin.read()
    n = int(input())
    print(fibonacci_sum_naive(n))

File: 01_week2/fibonacci_sum_last_digit/fibonacci_sum_last_digit.cpp
/*Task. Given an integer n, find the last digit of the sum F 0 + F 1 + · · · + F n .*/
#include <iostream>

int fibonacci_sum_naive(long long n) {
	if (n <= 1)
		return n;

	long long previous = 0;
	long long current  = 1;
	long long sum      = 1;

	for (long long i = 0; i < n - 1; ++i) {
		long long tmp_previous = previous;
		previous = current;
		current = tmp_previous %10 + current %10;
		sum += current;
	}

	return sum %10;

}

long long Fibonaccisumfast(long long n){
	if(n<=1)
		return n;
	long long previous=1;
	long long beforePrevious=0;
	long long current=-1;
	for(long long i=2;i<=n;i++){
		current=(previous+beforePrevious)%10;
	beforePrevious=previous;
	previous=current;
	}
	return current-1;
}

int main() {
	long long n = 0;
	std::cin >> n;
	//    std::cout << fibonacci_sum_naive(n)<<"\n";
	std::cout << Fibonaccisumfast((n+2)%60);
}
